### ISSUE TITLE
Make the max length of Textarea fields configurable

### ISSUE THREAD

**Issue by hannaofsweden** [ryancramerdesign/ProcessWire#1261]
Probably the biggest, and most easily fixable, limitation of Processwire is the rather arbitrary 16384 character limit imposed by the Sanitizer class on Textarea fields. This limit is much smaller than the database limit for the field type used (to be precise, about 1,000 times smaller, the mediumtext field type takes a maximum of 16,777,215 characters).

Removing this limitation would make Processwire suitable for handling large text dokuments, for instance, in a Wiki-like site or for a site that contains documentation or other long documents, at a negligible cost in terms of development time.

I suggest that the max length of Textarea fields be made configurable, with the current value as default and the maximum imposed by the database as its maximum. Since Textarea inherits from Text, which has a size option that Textarea removes, this should be very easy.

**Comment by LostKobrakai** [comment]
> (to be precise, about 1,000 times smaller, the mediumtext field type takes a maximum of 16,777,215 characters)

That's not correct. The 16 mio is the max. size in bytes, but the text is saved as multibyte string. UTF-8 contains chars, where chars need up to four bytes of space to be saved. It's still way more space than 16384 chars need, but also considerably less then the number you mentioned.

**Comment by hannaofsweden** [comment]
LostKobrakai: You're right, of course. Unicode characters are encoded with a variable number of bytes. So ideally the max length should be in bytes, not Unicode characters. The average for English is slightly above one byte per character since the most common characters are encoded with one byte. In Asian languages the average is three bytes per character, and in non-English scripts like Greek the average is about two.

Still, that doesn't change the point that users should be able to set a higher limit. For instance, you could limit it to a maximum of MySQL field size in bytes/3 characters. Or let the developer take responsibility for whatever changes he makes to the max size, and write a warning note saying "Warning: setting the max size too high may give database errors for extremely long texts".

However, this problem should rarely occur, because most text will be nowhere near the MySQL field limit, which should be somewhere close to 16 million charcters in English and five point something in, for instance, Chinese. I don't think many people will have texts that long, but some text may be something like 20 or 30 thousand characters.

Another possibillity is to make the limit editable, but only in the config file. That would make sure that only developers who know what they are doing increase the limit. Personally, though, I think it should be possible to change through the interface on the field level.

**Comment by teppokoivula** [comment]
> Personally, though, I think it should be possible to change through the interface on the field level.

Agreed. While my own needs for such a setting have mainly been for an opposite reason (imposing a much smaller character limit), this would be a very nice thing to have.

I don't think I've ever had a situation where the default character limit would've been too small (16k+ chars is quite a lot), but that's definitely possible too, and having this kind of limit hardcoded doesn't sound too good.

**Comment by ryancramerdesign** [comment]
Good points. I will make this configurable in this weeks commits.

**Comment by ryancramerdesign** [comment]
Okay just looking in the code now, I don't see that we're running FieldtypeTextarea or InputfieldTextarea through `$sanitizer->textarea()` for PW fields. It only applies that sanitizer if InputfieldTextarea is being used without a Fieldtype, like in the context of FormBuilder. So it should allow full use of any size supported by MySQL currently, as best as I can tell. Are you seeing it get truncated at 16384 characters somewhere? If so, I'm not sure where–we'll need to hunt it down.
